### PR TITLE
Migrate converter tests to be integration style.

### LINKED ISSUE #904: Switch each converter's tests from unit to integration
We should be testing these as installed in a `Retrofit` instance with `MockWebServer`.

### PR DESCRIPTION
**Pull request by JakeWharton**
This also fundamentally changes the `Type` passed to the converter. Instead of just passing the runtime class, we use the type parsed from the method declaration. This allows complex types like `List<Record<Person>>` to be handled by the converter instead of just receiving something like `ArrayList`.

Closes #904.

### CODE REVIEW

File: retrofit-converters/wire/pom.xml [hunk 1]
```diff
@@ -26,6 +26,12 @@
     <dependency>
       <groupId>com.squareup.wire</groupId>
       <artifactId>wire-runtime</artifactId>
+      <exclusions>
+        <exclusion>
+          <groupId>com.squareup.okio</groupId>
+          <artifactId>okio</artifactId>
+        </exclusion>
+      </exclusions>
```
Comment (JakeWharton) [inline retrofit-converters/wire/pom.xml]: This really should be needed, but for some reason the older Okio dependency from Wire's runtime is beating out the newer version which is specified by OkHttp (itself a transitive dependency of Retrofit).

File: retrofit-converters/jackson/src/test/java/retrofit/JacksonConverterTest.java [hunk 2]
```diff
@@ -1,89 +1,156 @@
...
-    private final String message;
-    private final int count;
+  @Rule public final MockWebServerRule server = new MockWebServerRule();
+
+  private Service service;
+
+  @Before public void setUp() {
+    SimpleModule module = new SimpleModule();
+    module.addSerializer(AnInterface.class, new AnInterfaceSerializer());
+    module.addDeserializer(AnInterface.class ,new AnInterfaceDeserializer());
```
Comment (swankjesse) [inline retrofit-converters/jackson/src/test/java/retrofit/JacksonConverterTest.java]: Not: comma

File: retrofit-converters/jackson/src/test/java/retrofit/JacksonConverterTest.java [hunk 3]
```diff
@@ -1,89 +1,156 @@
...
+    AnImplementation body = response.body();
+    assertThat(body.theName).isEqualTo("value");
 
-    @Override public int hashCode() {
-      int result = message != null ? message.hashCode() : 0;
-      result = 31 * result + count;
-      return result;
-    }
+    RecordedRequest request = server.takeRequest();
+    // TODO figure out how to get Jackson to stop using AnInterface's serializer here.
```
Comment (swankjesse) [inline retrofit-converters/jackson/src/test/java/retrofit/JacksonConverterTest.java]: Bizarre.

File: retrofit-converters/moshi/src/main/java/retrofit/MoshiConverter.java [hunk 4]
```diff
@@ -38,7 +38,7 @@ public MoshiConverter() {
   public MoshiConverter(Moshi moshi) {
     if (moshi == null) throw new NullPointerException("moshi == null");
     this.moshi = moshi;
-    this.mediaType = MediaType.parse("application/json; charset=utf-8");
+    this.mediaType = MediaType.parse("application/json; charset=UTF-8");
```
Comment (swankjesse) [inline retrofit-converters/moshi/src/main/java/retrofit/MoshiConverter.java]: (Probably worthwhile to grep and make sure 100% of your projects are consistent.)

File: retrofit-converters/moshi/src/test/java/retrofit/MoshiConverterTest.java [hunk 5]
```diff
@@ -51,45 +49,75 @@
...
+      jsonWriter.endObject();
     }
 
-    @FromJson public Example from(JsonReader reader) throws IOException {
-      throw new UnsupportedOperationException(); // Moshi requires this method to exist.
+    @FromJson public AnInterface read(JsonReader jsonReader) throws IOException {
+      jsonReader.beginObject();
+
+      String name = null;
+      while (jsonReader.peek() != JsonReader.Token.END_OBJECT) {
```
Comment (swankjesse) [inline retrofit-converters/moshi/src/test/java/retrofit/MoshiConverterTest.java]: `while (jsonReader.hasNext())`

... Unless you're going for symmetry with the other libraries!

File: retrofit-converters/protobuf/src/test/java/retrofit/ProtoConverterTest.java [hunk 6]
```diff
@@ -1,73 +1,119 @@
...
-  @Test public void serialize() throws Exception {
-    RequestBody body = converter.toBody(PROTO, Phone.class);
-    assertThat(body.contentType().toString()).isEqualTo("application/x-protobuf");
-    assertBody(body).isEqualTo(ENCODED_PROTO);
+  @Before public void setUp() {
+    Retrofit retrofit = new Retrofit.Builder()
+        .endpoint(server.getUrl("/").toString())
+        .converter(new ProtoConverter())
+        .build();
+    service = retrofit.create(Service.class);
```
Comment (swankjesse) [inline retrofit-converters/protobuf/src/test/java/retrofit/ProtoConverterTest.java]: The converter factory also let's you fast fail here.

File: retrofit-converters/protobuf/src/test/java/retrofit/ProtoConverterTest.java [hunk 7]
```diff
@@ -1,73 +1,119 @@
...
+    assertThat(request.getHeader("Content-Type")).isEqualTo("application/x-protobuf");
+  }
+
+  @Test public void empty() throws IOException {
+    server.enqueue(new MockResponse());
+
+    Call<Phone> call = service.get();
+    Response<Phone> response = call.execute();
+    Phone body = response.body();
+    assertThat(body.hasNumber()).isFalse();
```
Comment (swankjesse) [inline retrofit-converters/protobuf/src/test/java/retrofit/ProtoConverterTest.java]: Proto encoding is weird.

File: retrofit-converters/protobuf/src/test/java/retrofit/ProtoConverterTest.java [hunk 8]
```diff
@@ -1,73 +1,119 @@
...
+    server.enqueue(new MockResponse().setBody(new Buffer().write(encoded)));
+
+    Call<?> call = service.get();
     try {
-      converter.fromBody(protoResponse("////"), Phone.class);
+      call.execute();
       fail();
-    } catch (RuntimeException expected) {
-      assertThat(expected.getCause() instanceof InvalidProtocolBufferException);
+    } catch (RuntimeException e) {
```
Comment (swankjesse) [inline retrofit-converters/protobuf/src/test/java/retrofit/ProtoConverterTest.java]: Can we be more specific than RuntimeException in what we their here? Moshi's JsonDataException seems like a usability win.

File: retrofit-converters/protobuf/src/test/java/retrofit/ProtoConverterTest.java (hunk 7 above)
Comment (JakeWharton) [inline retrofit-converters/protobuf/src/test/java/retrofit/ProtoConverterTest.java]: Yep. We were erroneously fast-tracking an empty body to `null` in 1.x which broke this use case.

File: retrofit-converters/protobuf/src/test/java/retrofit/ProtoConverterTest.java (hunk 6 above)
Comment (JakeWharton) [inline retrofit-converters/protobuf/src/test/java/retrofit/ProtoConverterTest.java]: :+1:

File: retrofit-converters/simplexml/src/test/java/retrofit/MyObject.java [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+  @Override public int hashCode() {
+    final int prime = 31;
+    int result = 1;
+    result = prime * result + count;
+    result = prime * result + ((message == null) ? 0 : message.hashCode());
+    return result;
+  }
+
+  @Override public boolean equals(Object obj) {
+    if (this == obj) {
```
Comment (swankjesse) [inline retrofit-converters/simplexml/src/test/java/retrofit/MyObject.java]: who writes code like this?

File: retrofit-converters/simplexml/src/test/java/retrofit/MyObject.java [hunk 10]
```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright (C) 2013 Square, Inc.
```
Comment (swankjesse) [inline retrofit-converters/simplexml/src/test/java/retrofit/MyObject.java]: moved file or mistake?

File: retrofit-converters/protobuf/src/test/java/retrofit/ProtoConverterTest.java [hunk 11]
```diff
@@ -1,73 +1,119 @@
...
+        .endpoint(server.getUrl("/").toString())
+        .converter(new ProtoConverter())
+        .build();
+    service = retrofit.create(Service.class);
   }
 
-  @Test public void deserialize() throws Exception {
-    Object proto = converter.fromBody(protoResponse(ENCODED_PROTO), Phone.class);
-    assertThat(proto).isEqualTo(PROTO);
+  @Test public void bodyWays() throws IOException, InterruptedException {
```
Comment (swankjesse) [inline retrofit-converters/protobuf/src/test/java/retrofit/ProtoConverterTest.java]: bothWays ?

File: retrofit/src/main/java/retrofit/RequestBuilder.java [hunk 12]
```diff
@@ -317,10 +320,11 @@ void setArguments(Object[] args) {
         if (value == null) {
           throw new IllegalArgumentException("Body parameter value must not be null.");
         }
-        if (value instanceof RequestBody) {
+        if (requestType == RequestBody.class
+            || (requestType == Object.class && value instanceof RequestBody)) {
           body = (RequestBody) value;
         } else {
-          body = converter.toBody(value, value.getClass());
+          body = converter.toBody(value, requestType);
```
Comment (swankjesse) [inline retrofit/src/main/java/retrofit/RequestBuilder.java]: win

**Comment by swankjesse** [comment]
Dig.

The PR description might also mention that there's a behavior change!

**Comment by swankjesse** [comment]
LGTM

File: retrofit-converters/simplexml/src/test/java/retrofit/MyObject.java (hunk 10 above)
Comment (JakeWharton) [inline retrofit-converters/simplexml/src/test/java/retrofit/MyObject.java]: nested class -> top-level class

File: retrofit-converters/simplexml/src/test/java/retrofit/MyObject.java (hunk 9 above)
Comment (JakeWharton) [inline retrofit-converters/simplexml/src/test/java/retrofit/MyObject.java]: Hah. IDEA probably generated this craziness.

File: retrofit-converters/protobuf/src/test/java/retrofit/ProtoConverterTest.java (hunk 8 above)
Comment (JakeWharton) [inline retrofit-converters/protobuf/src/test/java/retrofit/ProtoConverterTest.java]: Follow-up. There's also problems with Wire here.

**Pull request merged**